modules/apigee_edge_teams/src/TeamContextManager.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge_teams;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Url;
+
+/**
+ * Describes the `apigee_edge_teams.context_manager` service.
+ *
+ * This service is responsible for understanding the context of the current
+ * route. Context will either be developer context or team context.
+ */
+class TeamContextManager implements TeamContextManagerInterface {
+
+  /**
+   * The current route match.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * TeamContextManager constructor.
+   *
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match.
+   */
+  public function __construct(RouteMatchInterface $route_match) {
+    $this->routeMatch = $route_match;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDestinationUrlForEntity(EntityInterface $entity): ?Url {
+    if ($corresponding_route_name = $this->getCorrespondingRouteNameForEntity($entity)) {
+      $parameters = $this->routeMatch->getParameters();
+      $parameters->replace([
+        $entity->getEntityTypeId() => $entity->id(),
+      ]);
+      return Url::fromRoute($corresponding_route_name, $parameters->all());
+    }
+
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCorrespondingRouteNameForEntity(EntityInterface $entity): ?string {
+    if ($current_route_object = $this->routeMatch->getRouteObject()) {
+      // If the route has only one and the same parameter, return the current
+      // route.
+      $parameters = $this->routeMatch->getRawParameters();
+      if ($parameters->has($entity->getEntityTypeId()) && $parameters->count() === 1) {
```

Comment:
This is interesting. I see the problem with the corresponding route option is that it doesn't tell you if you should be showing the switcher to switch teams when there is no corresponding route to change to a developer route.
Some routes do have the `team` parameter but aren't compatible with other teams (such as `entity.subscription.team`). But that route has more than one parameter. I think this will work given the current state of things but I do worry about the switcher showing up where it isn't wanted because it is using the `{team}` parameter as it's only parameter in a custom implementation. I would almost rather have a static list of routes that is source of truth as to where the switcher shows up. for instance, I see the switcher shows up on `user/{uid}/orders` when the commerce module is installed. The team links fall back to their canonical links. We should see if this fallback is needed.

Reply:
Yeah. Right now if we can build a URL from existing parameters and user has access to this URL, the switcher will be displayed and other entities fallback to their canonical URL. This is more a UX thing. On the legacy site, the switcher is available on all pages. It seems weird (bad UX) to have it on some pages only.

Let me know how to proceed. @Jaesin @cnovak